### DISCUSSION TITLE
What is the difference between using `InferenceRPUConfig` vs any presets for inference?

### DISCUSSION THREAD

**Discussion by matifali** [IBM/aihwkit#435] [Q&A] [answered]
What is the difference between using `InferenceRPUConfig` vs any presets to be used for inference? e.g.

```python
# 1st case
RPU_CONFIG = InferenceRPUConfig()
RPU_CONFIG.backward.bound_management = BoundManagementType.NONE
RPU_CONFIG.forward.out_res = 2 / (2**8 - 1)
RPU_CONFIG.forward.inp_res = 2 / (2**8 - 1)
RPU_CONFIG.forward.w_noise_type = WeightNoiseType.ADDITIVE_CONSTANT
RPU_CONFIG.forward.w_noise = W_NOISE
RPU_CONFIG.forward.out_noise = O_NOISE
RPU_CONFIG.noise_model = PCMLikeNoiseModel(g_max=G_MAX)
# 2nd case
RPU_CONFIG = ReRamSBPreset()
# 3rd case
RPU_CONFIG = PCMPreset()
```
I have pre-trained models that I am converting to analog using
I am getting a way better accuracy if I use `InferenceRPUConfig` with `PCMLikeNoiseModel`
but very bad performance with `ReRamSBPreset` and `PCMPreset`

What is the difference between using an `InferenceRPUConfig` with `PCMLikeNoiseModel` and a `PCMPreset`? How does the `drift_analog_weights` for both cases?

**Comment by maljoras** [comment, marked as answer]
The `InferenceRPUConfig` is used for accuracy estimation for a situation where the resistive crossbar arrays are only used in the inference pass (training is done in FP and only inference is using the resistive devices). The other presets are used for in-memory training with forward / backward and update in "analog" and updating using stochastic pulse trains in a fully parallel manner. This is typically very challenging and might need special optimizers such as `MixedPrecision` / or `TikiTaka` for which we also have presets.    

For inference only chips, the expected programming noise and drift can be set by the `noise_model` setting.

**Reply by matifali** [reply]
Thank you, @maljoras. Can we set `d2d` and `c2c` variance for the inference-only cases?
Also, is a noise model for `ReRamSBPreset` to be used only in inference?

**Reply by maljoras** [reply]
`ReRamSBPreset` determines the pulse response of the conductance. It is typically only used for training. However, of course you could try to explicitly program the inference weights using the update and a read-write-verify scheme. 

Device-to-device variation for inference is implicitly controlled by setting the programming noise [polynomials coefficients](https://aihwkit.readthedocs.io/en/latest/api/aihwkit.inference.noise.pcm.html?highlight=pcmlikenoise#aihwkit.inference.noise.pcm.PCMLikeNoiseModel) (`noise_model.prog_coeff`). See [here](https://aihwkit.readthedocs.io/en/latest/pcm_inference.html#programming-noise) for more details of the noise model.
